pkg/src/src/main/scala/edu/berkeley/cs/amplab/sparkr/SerializeJavaR.scala

```diff
@@ -0,0 +1,297 @@
+package edu.berkeley.cs.amplab.sparkr
+
+import scala.collection.mutable.HashMap
+import scala.collection.JavaConversions._
+
+import java.io.DataInputStream
+import java.io.DataOutputStream
+
+/**
+ * Utility functions to serialize, deserialize objects to / from R
+ */
+object SerializeJavaR {
+
+  // Type mapping from R to Java
+  //
+  // integer -> Int
+  // character -> String
+  // logical -> Boolean
+  // double, numeric -> Double
+  // raw -> Array[Byte]
+  //
+  // list[T] -> Array[T], where T is one of above mentioned types
+  // environment -> Map[String, T], where T is a native type
+  // jobj -> Object, where jobj is an object created in the backend
+
+  def readObjectType(dis: DataInputStream) = {
+    dis.readByte().toChar
+  }
+
+  def readObject(dis: DataInputStream): Object = {
+    val dataType = readObjectType(dis)
+    readTypedObject(dis, dataType)
+  }
+
+  def readTypedObject(
+      dis: DataInputStream,
+      dataType: Char): Object = {
+    dataType match {
+      case 'i' => new java.lang.Integer(readInt(dis))
+      case 'd' => new java.lang.Double(readDouble(dis))
+      case 'b' => new java.lang.Boolean(readBoolean(dis))
+      case 'c' => readString(dis)
+      case 'e' => readMap(dis)
+      case 'r' => readBytes(dis)
+      case 'l' => readList(dis)
+      case 'j' => JavaObjectTracker.getObject(readString(dis))
+      case _ => throw new IllegalArgumentException(s"Invalid type $dataType")
+    }
+  }
+
+  def readBytes(in: DataInputStream) = {
+    val len = readInt(in)
+    val out = new Array[Byte](len)
+    val bytesRead = in.read(out, 0, len)
+    assert(len == bytesRead)
+    out
+  }
+
+  def readInt(in: DataInputStream) = {
+    in.readInt()
+  }
+
+  def readDouble(in: DataInputStream) = {
+    in.readDouble()
+  }
+
+  def readString(in: DataInputStream) = {
+    val len = in.readInt()
+    val asciiBytes = new Array[Byte](len)
+    in.read(asciiBytes, 0, len)
+    assert(asciiBytes(len - 1) == 0)
+    val str = new String(asciiBytes.dropRight(1).map(_.toChar))
+    str
+  }
+
+  def readBoolean(in: DataInputStream) = {
+    val intVal = in.readInt()
+    if (intVal == 0) false else true
+  }
+
+  def readBytesArr(in: DataInputStream) = {
+    val len = readInt(in)
+    (0 until len).map(_ => readBytes(in)).toArray
```

Comment:
Just a note: this might be pretty slow compared to while loops. I guess for now it's okay to leave it as is.